### DISCUSSION TITLE
Support data objects

### DISCUSSION THREAD

**Discussion by jeffreyvanhees** [dedoc/scramble#168] [Ideas]
For automatic generation of typescript files, I use [spatie/laravel-data](https://github.com/spatie/laravel-data/) in combination with [spatie/typescript-transformer](https://github.com/spatie/typescript-transformer) instead of default resources. It would be great if this is supported for generating schemas, example responses and such. What do you think about it?

**Comment by romalytvynenko** [comment]
@jeffreyvanhees it would be great indeed. 

Can you please share some examples of your code of how you use these libs for requests and responses?

**Reply by jeffreyvanhees** [reply]
I don't really have any good examples yet, but from their documentation it should be possible to build support for this.

On this page it's very well explained how data objects from request are working:
https://spatie.be/docs/laravel-data/v3/as-a-data-transfer-object/request-to-data-object

... and also how requests can be validated:
https://spatie.be/docs/laravel-data/v3/as-a-data-transfer-object/request-to-data-object#content-validating-a-collection-of-data-objects

And here; how it transforms to a JSON-resource:
https://spatie.be/docs/laravel-data/v3/as-a-resource/from-data-to-resource

When using [spatie/typescript-transformer](https://github.com/spatie/typescript-transformer), you'll need to add the typescript-annotation to the data object and the package will generate a typescript file out of the box:

Example:

**App\Data\ProductData.php**
```php
namespace App\Data;

#[Typescript]
class ProductData extends Data
{
    public function __construct(
        
        public ?int $id,
    
        #[Max(100)]
        public string $name,
        
        #[Max(200)]
        public string $description,
    ) {
        //
    }
}
```

**App\Http\Controllers\ProductController.php**
```php
namespace App\Http\Controllers;

use App\Data\ProductData;
use App\Models\Product;

class ProductController extends Controller 
{
    // Now it's used like a Resource
    public function index() 
    {
        $products = Product::all();
        
        return ProductData::collection($products);
    }
    
    // Now it's used like a FormRequest
    public function store(ProductData $productData) 
    {
        Product::create($productData->all());
        
        return redirect()->back();
    }
    
    // Now it's used like a FormRequest
    public function update(Product $product, ProductData $productData) 
    {
        Product::update($productData->all());
        
        return redirect()->back();
    }
}
```
  

After running `php artisan typescript:transform`:
**resources/types/types.d.ts**
```ts
// ...
class ProductData
{
    public ?int $id;
    public string $name;
    public string $description;
} 
// ...
```

**Comment by jeffreyvanhees** [comment]
Any thoughts on this @romalytvynenko? :)

**Reply by romalytvynenko** [reply]
@jeffreyvanhees sorry for 3 moths reply delay. I love the idea. Definitely will work on this. Not sure when. PRs/packages are welcome.

**Reply by jeffreyvanhees** [reply]
No problem @romalytvynenko! Keep up the good work!

**Comment by romalytvynenko** [comment]
Hey @jeffreyvanhees 

I will definitely a support for `spatie/laravel-data` package.

Just cannot tell you when exactly that would happen 🙃

**Reply by asterd** [reply]
Hi! any update on this?

**Reply by romalytvynenko** [reply]
@asterd no updates on this for now, still planned.

**Comment by igorsgm** [comment]
Hello @romalytvynenko 
Just checking if you've made any progress on this.

This package is amazing.
Thanks!

**Comment by jeffreyvanhees** [comment]
Now that this is available in Scramble Pro, I think it's ok to close this discussion.
